ntoskrnl/ke/amd64/trap.S

```diff
@@ -1140,19 +1132,33 @@ PUBLIC KiSetTrapContext
  */
 EXTERN KiDeliverApc:PROC
 
+/*
+ * VOID
+ * KiInitiateUserApc(
+ *     _In_ PKTRAP_FRAME TrapFrame@<rcx>);
+ *
+ * This function is called to deliver user mode APCs.
+ * It clobbers all non-volatile registers, except rax.
+ */
 PUBLIC KiInitiateUserApc
 .PROC KiInitiateUserApc
 
     /* Generate a KEXCEPTION_FRAME on the stack */
     GENERATE_EXCEPTION_FRAME
 
+    /* Save rax to not clobber the return for the system call handler */
+    mov [rsp + ExP1Home], rax
+
     /* Raise IRQL to APC_LEVEL */
     mov rax, APC_LEVEL
     mov cr8, rax
 
     /* Get the current thread */
     mov rbp, gs:[PcCurrentThread]
```

Comment:
i see elsewhere you've changed to `gs:qword ptr [PcCurrentThread]`

Reply:
It's not actually needed. Changed it to `gs:[PcCurrentThread]` now.